### PR TITLE
Ultra PSET Spacecraft Frame + PSET Helio Frame Exposure Times

### LINKED ISSUE #1039: Ultra PSET Spacecraft Frame + PSET Helio Frame Exposure Times
### Summary

_No response_

### Requirements

The more specific the better.
  - [ ]
  - [ ]

### Related tickets

_No response_

### PR DESCRIPTION
**Pull request by laspsandoval**
# Change Summary

## Overview
Finished pset spacecraft frame EXCEPT background rates
Added pset helio frame exposure times

## New Files
- dps_sensitivity45.cdf
   - Lookup table
   
## Updated Files
- ultra_l1c_pset_bins.py
   - Finished pset spacecraft frame EXCEPT background rates
   - Added pset helio frame exposure times
- conftest.py
   - Modified path 
- constants.py
   - Added constants
   
## Testing
- test_ultra_l1c_pset_bins.py
   - Tests ultra_l1c_pset_bins.py
- Test data added: dps_exposure_helio_45_E1.cdf, dps_exposure_helio_45_E12.cdf, dps_exposure_helio_45_E24.cdf

### CODE REVIEW

**Comment by laspsandoval** [comment]
pre-commit.ci autofix

**Comment by demajistre** [comment]
There may be a problem merging this back into the dev branch because of unsynchronized changes to the _decom.py_ file. In the PR branch, line 11 of this file is:
`from space_packet_parser import parser, xtcedef`
whereas on the dev branch it is
`from space_packet_parser import definitions`

The dev branch code is more recent and is part of a more comprehensive change that upgrades  _space_packet_parser_ from version 4 to version 5. It looks like this PR was branched before these changes were made to dev.

The bottom line here is that the code in this PR will not run with version 5 of the _space_packet_parser_, i.e., this code won't run with the package configuration specified by the _poetry.lock_ file currently in dev.

This may be as simple as changing _decom.py_ to the new version in dev, but I really have no idea what will happen if you try to merge it back in.

File: imap_processing/ultra/l1c/ultra_l1c_pset_bins.py [hunk 1]
```diff
@@ -174,3 +209,123 @@ def get_pointing_frame_exposure_times(
...
+        # Normalized vectors representing the direction of the heliocentric velocity.
+        helio_normalized = helio_velocity.T / np.linalg.norm(
+            helio_velocity.T, axis=1, keepdims=True
+        )
+        # Converts vectors from Cartesian coordinates (x, y, z)
+        # into spherical coordinates
+        az, el, _ = cartesian_to_spherical(-helio_normalized)
+
+        # Bin the coordinates.
+        az_idx = np.digitize(az, az_bin_edges) - 1
```
Comment (sdhoyt) [inline imap_processing/ultra/l1c/ultra_l1c_pset_bins.py:291]: This is cool. I've never seen this function before

File: imap_processing/ultra/l1c/ultra_l1c_pset_bins.py [hunk 2]
```diff
@@ -112,6 +120,33 @@ def cartesian_to_spherical(
     return np.degrees(az), np.degrees(el), magnitude_v
 
 
+def spherical_to_cartesian(
```
Comment (sdhoyt) [inline imap_processing/ultra/l1c/ultra_l1c_pset_bins.py:123]: I wonder if it would be good to pull this out into a common utility file. Not sure if others will need this though.

**Review by sdhoyt — APPROVED** [review]
The code looks good to me!

File: imap_processing/tests/ultra/unit/conftest.py [hunk 3]
```diff
@@ -176,14 +140,7 @@ def events_fsw_comparison_theta_0():
...
     )
-    return (
-        Path(sys.modules[__name__.split(".")[0]].__file__).parent
-        / "tests"
-        / "ultra"
-        / "test_data"
-        / "l0"
-        / filename
-    )
+    return imap_module_directory / "tests" / "ultra" / "test_data" / "l0" / filename
```
Comment (tmplummer) [inline imap_processing/tests/ultra/unit/conftest.py:143]: In the vein of reducing duplicated code and making refactors easier, I would suggest creating a fixture pointing to each of the following directories that build upon each other: `ultra_test_data`, `ultra_l0_test_data`, `ultra_l1_test_data`, etc. This is just personal preference, but makes it so that if you change a directory name for some reason, there is just one change needed to fix all the tests that would be broken by that change. For example see here: https://github.com/IMAP-Science-Operations-Center/imap_processing/blob/dev/imap_processing/tests/hi/conftest.py#L11

File: imap_processing/ultra/l1c/ultra_l1c_pset_bins.py (hunk 2 above)
Comment (tmplummer) [reply imap_processing/ultra/l1c/ultra_l1c_pset_bins.py:123]: +1. This could go in the `spice.geometry` module. This move could be done in a separate PR. If/when that refactor is done, I would suggest that the function signature should be:
```
def sphreical_to_cartesian(spherical_coords: np.ndarray) -> np.ndarray:
```
Where both input and output coordinates follow:  For a single coordinate, shape is (3,) for multiple coordinates, shape is (n, 3).

File: imap_processing/ultra/l1c/ultra_l1c_pset_bins.py [hunk 4]
```diff
@@ -174,3 +209,123 @@ def get_pointing_frame_exposure_times(
...
+    r = np.ones(el_grid.shape)
+    x, y, z = spherical_to_cartesian(r, np.radians(az_grid), np.radians(el_grid))
+
+    # Reshape and combine the Cartesian coordinates into a 2D array.
+    cartesian = np.vstack(
+        [x.flatten(order="F"), y.flatten(order="F"), z.flatten(order="F")]
+    )
+
+    # Spacecraft velocity in the pointing (DPS) frame wrt heliosphere.
+    state, lt = spice.spkezr("IMAP", time, "IMAP_DPS", "NONE", "SUN")
```
Comment (tmplummer) [inline imap_processing/ultra/l1c/ultra_l1c_pset_bins.py:263]: There is a common function for this: https://github.com/IMAP-Science-Operations-Center/imap_processing/blob/dev/imap_processing/spice/geometry.py#L86
```suggestion
    state = imap_state(time, SpiceFrame.IMAP_DPS, SpiceBody.Sun)
```

File: imap_processing/ultra/l1c/ultra_l1c_pset_bins.py [hunk 5]
```diff
@@ -174,3 +209,123 @@ def get_pointing_frame_exposure_times(
...
+    time: np.ndarray,
+    sc_exposure: np.ndarray,
+) -> np.ndarray:
+    """
+    Compute a 3D array of the exposure in the helio frame.
+
+    Parameters
+    ----------
+    time : np.ndarray
+        Median time of pointing.
```
Comment (tmplummer) [inline imap_processing/ultra/l1c/ultra_l1c_pset_bins.py:226]: It looks like this is being used by spice so this should be specified to be `J2000 seconds`.

File: imap_processing/ultra/l1c/ultra_l1c_pset_bins.py [hunk 6]
```diff
@@ -174,3 +209,123 @@ def get_pointing_frame_exposure_times(
...
+
+    # Create a 3D Cartesian grid from spherical coordinates
+    # using azimuth and elevation midpoints.
+    az_grid, el_grid = np.meshgrid(az_bin_midpoints, el_bin_midpoints[::-1])
+
+    # Radial distance.
+    r = np.ones(el_grid.shape)
+    x, y, z = spherical_to_cartesian(r, np.radians(az_grid), np.radians(el_grid))
+
+    # Reshape and combine the Cartesian coordinates into a 2D array.
```
Comment (tmplummer) [inline imap_processing/ultra/l1c/ultra_l1c_pset_bins.py:257]:
```suggestion
    # Reshape and combine the Cartesian coordinates into a 3D array.
```

File: imap_processing/ultra/l1c/ultra_l1c_pset_bins.py [hunk 7]
```diff
@@ -174,3 +209,123 @@ def get_pointing_frame_exposure_times(
...
+    cartesian = np.vstack(
```
Comment (tmplummer) [inline imap_processing/ultra/l1c/ultra_l1c_pset_bins.py:258]: I personally think that three-dimensional coordinates should generally be expressed as `n x 3` arrays for `n > 1` and 3 element vectors for a single coordinate. Modifying the `spherical_to_cartesian` function to behave this way would eliminate the need to reshape the output like is being done here.

File: imap_processing/ultra/l1c/ultra_l1c_pset_bins.py [hunk 8]
```diff
@@ -174,3 +209,123 @@ def get_pointing_frame_exposure_times(
...
+
+        # Normalized vectors representing the direction of the heliocentric velocity.
+        helio_normalized = helio_velocity.T / np.linalg.norm(
+            helio_velocity.T, axis=1, keepdims=True
+        )
+        # Converts vectors from Cartesian coordinates (x, y, z)
+        # into spherical coordinates
+        az, el, _ = cartesian_to_spherical(-helio_normalized)
+
+        # Bin the coordinates.
```
Comment (tmplummer) [inline imap_processing/ultra/l1c/ultra_l1c_pset_bins.py:290]: I think that this binning exercise is once again just a 2D histogram, no?

File: imap_processing/ultra/l1c/ultra_l1c_pset_bins.py [hunk 9]
```diff
@@ -174,3 +209,123 @@ def get_pointing_frame_exposure_times(
...
+        # to every position on the grid in the despun grid.
+        helio_velocity = spacecraft_velocity.reshape(3, 1) + energy_velocity * cartesian
+
+        # Normalized vectors representing the direction of the heliocentric velocity.
+        helio_normalized = helio_velocity.T / np.linalg.norm(
+            helio_velocity.T, axis=1, keepdims=True
+        )
+        # Converts vectors from Cartesian coordinates (x, y, z)
+        # into spherical coordinates
+        az, el, _ = cartesian_to_spherical(-helio_normalized)
```
Comment (tmplummer) [inline imap_processing/ultra/l1c/ultra_l1c_pset_bins.py:288]: A comment on why the negative sign here would be good. I think I understand it as `helio_normalized` is the mean ENA particle velocity vector, so you have to negate it to get the look direction?

File: imap_processing/ultra/l1c/ultra_l1c_pset_bins.py [hunk 10]
```diff
@@ -174,3 +209,123 @@ def get_pointing_frame_exposure_times(
         exposure = cdf_file.varget(f"dps_grid{sensor}") * n_spins
 
     return exposure
+
+
+@ensure_spice
+@typing.no_type_check
```
Comment (tmplummer) [inline imap_processing/ultra/l1c/ultra_l1c_pset_bins.py:215]: If you use the common `imap_state` function as commented about below, these decorators can be removed.

File: imap_processing/tests/ultra/unit/test_ultra_l1c_pset_bins.py [hunk 11]
```diff
@@ -27,15 +32,34 @@ def test_data():
     return v, energy
 
 
+@pytest.fixture()
+def kernels(spice_test_data_path):
+    """List SPICE kernels."""
+    required_kernels = [
+        "imap_science_0001.tf",
```
Comment (tmplummer) [inline imap_processing/tests/ultra/unit/test_ultra_l1c_pset_bins.py:39]: It looks like the only difference between this list of kernels and the `imap_ena_sim_metakernel.template` metakernel is this kernel which should probably be added to the metakernel. I would recommend using that or the fixture [here](https://github.com/IMAP-Science-Operations-Center/imap_processing/blob/dev/imap_processing/tests/conftest.py#L143) to furnish an explicit list of kernels.

**Review by tmplummer — COMMENTED** [review]
Looks like I was too late. I have lots of comments on this PR though.
